Add pointerrest and pointerrestout events to map component

marcjansen opened the pull request:
This PR suggest adding two new events to `GeoExt.component.Map`:  `pointerrest` and `pointerrestout`:

With these events it easy to create nice hovering effects, that do not fire as often as the raw pointermove does.

![pointerrest2](https://cloud.githubusercontent.com/assets/227934/8302440/99bdb16a-1996-11e5-9e94-11202e4a5ea1.png)

The Popup example has been changed to now work with these new events.

Additionally some API docs were added to certain Map-properties.

Please review.

marcjansen commented:
Implementation heavily inspired by @dnlkoch

marcjansen commented:
Alternatively this may be turned into a plugin, if others would prefer this.

bentrm commented:
I think this are very useful additions, I just have some questions:
* Would it also be possible to fire these events as native ol.MapEvents of the ol.Map that are added by the GeoExt.component.Map on instantiation or via plugin? As a user, I would expect to listen on the map for this as with other map events? Maybe we can provide a set of plugins that add theses events to the ol.Map?
* Is `pointerrestintolerance` strictly needed by your implementation? I would favor to set a switch like `hasBeenResting` on the `pointerrest` events parameters to check if the pointer has been resting inside of the tolerance bounds before. (I even would have expected to check if the pointer has been resting before by listening to `pointerout` to proof the opposite, though I'm not 100% sure.)
* I think `pointerout` is ill named. My advice would be `pointerunrest` or at least `pointerrestout`.
* If `pointerrestintolerance` is needed I would like to apply for the beer with `pointerpartialunrest`.
Anyway, I can think of many uses for this, so great work!

marcjansen commented:
> * Would it also be possible to fire these events as native ol.MapEvents of the ol.Map that are added by the GeoExt.component.Map on instantiation or via plugin? As a user, I would expect to listen on the map for this as with other map events? Maybe we can provide a set of plugins that add theses events to the ol.Map?

I don think this can be done with the compiled version of ol, and I don't think we schould do it. AFAIK `dispatchEvent` is not exported and throwing events on objects you don't *own* feels wrong to me.


> * Is pointerrestintolerance strictly needed by your implementation? I would favor to set a switch like hasBeenResting on the pointerrest events parameters to check if the pointer has been resting inside of the tolerance bounds before. (I even would have expected to check if the pointer has been resting before by listening to pointerout to proof the opposite, though I'm not 100% sure.)

You are right `pointerrestintolerance` just happened to be one code path which I do not need, but which may prove useful for someone else. An additional argument for pointerrest may also be nice, but I like the explicitly named event better. Are your feelings strong against `pointerrestintolerance`? If so I can rework this.

> * I think pointerout is ill named. My advice would be pointerunrest or at least pointerrestout

OK, agreed, will update to make this `pointerrestout`.

> * If pointerrestintolerance is needed I would like to apply for the beer with pointerpartialunrest.

That is even worse IMHO :stuck_out_tongue_winking_eye:

Let't keep this up for some time, maybe someone has an idea...

Thanks for the feedback, @bentrm

marcjansen commented:
Hey @bentrm, I changed my mind wrt `pointerrestintolerance` and removed it.

Now in that codepath I simply fire a `pointerrest`, with the original location of the first `pointerrest`. I definitely like this better, and you now officially qualify for a drink on me :beers:

KaiVolland commented:
Looks good! Thanks @marcjansen @dnlkoch @bentrm

marcjansen commented:
Cheers for the review, @KaiVolland
